onnx/defs/tensor/defs.cc

```diff
@@ -89,7 +89,98 @@ ONNX_OPERATOR_SET_SCHEMA(
+          // the dimensions we are setting outputShape in the outputProduct variable.
+          // The outputProduct will potentially be used for inferring a dimension marked -1.
+          auto* outputShape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          TensorShapeProto::Dimension* negativeOneDim = nullptr;
+          const auto& dataInputTensorType = ctx.getInputType(0)->tensor_type();
+          bool unresolvedZero = false;
+          int64_t outputProduct = 1;
+          for (int i = 0; i < static_cast<int>(targetShape.size()); ++i)  {
+            // Add a new dimension to outputShape
+            auto* new_dim = outputShape->add_dim();    
+            if (targetShape[i] == -1) {
+              // Check if multiple -1's. If not, set negativeOneDim, marking this
+              // dimension to potentially be filled in later.
+              if (negativeOneDim) {
+                fail_shape_inference("Target shape may not have multiple -1 dimensions");
+              }
+              negativeOneDim = new_dim;
+            } else if (targetShape[i] == 0) {
+              // Check if data input has a shape, if the index i is within its bounds, and
+              // the dimension value is present at index i. If these conditions are satisfied,
+              // the dimension value can be inferred and set. Otherwise, set unresolvedZero flag.
+              if (dataInputTensorType.has_shape() && 
+                  i < dataInputTensorType.shape().dim_size() && 
+                  dataInputTensorType.shape().dim(i).has_dim_value()) {
+                const auto& dim_value = dataInputTensorType.shape().dim(i).dim_value();
+                new_dim->set_dim_value(dim_value);
+                outputProduct *= dim_value;
+              } else {
+                unresolvedZero = true;
+              }
+            } else if (targetShape[i] > 0) {
+              // Set the dimension value to targetShape[i]
+              new_dim->set_dim_value(targetShape[i]);
+              outputProduct *= targetShape[i];
+            } else {
+              // Check if value is less than -1; fail if so
+              fail_shape_inference("Target shape may not have dimensions less than -1");
+            }
+          }
+
+          // If negativeOneDim has been set, we attempt to infer its value. This can be done if 
+          // all other dimensions have been set (unresolvedZero is false) and complete shape
+          // information is present for the data input tensor.
+          if (negativeOneDim && !unresolvedZero) {
+            // First, attempt to compute number of elements in input data tensor, setting flag
+            // to false if not possible
+            int64_t dataElements = 1;
+            bool dataElementsFlag = true;
+            if (!dataInputTensorType.has_shape()) {
+              dataElementsFlag = false;
+            } else {
+              for (int i = 0; i < dataInputTensorType.shape().dim_size(); ++i)  {
```

Comment:
nit: probably can do 
```
for (const auto& dim : dataInputTensorType.shape.dim()) {
...
}
```

Reply:
Good point, but now I actually use the i